### DISCUSSION TITLE
Global API

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#187] [Guide]
Here are the global api methods and their types. 
I'll add more explanation/organization soooooon :D

> Note: These are all proxied/lazy loaded. They won't be "required" until you actually invoke the method.

```js
cd: typeof shelljs.cd
cp: typeof shelljs.cp
chmod: typeof shelljs.chmod
echo: typeof shelljs.echo
exec: typeof shelljs.exec
exit: typeof shelljs.exit
grep: typeof shelljs.grep
ln: typeof shelljs.ln
ls: typeof shelljs.ls
mkdir: typeof shelljs.mkdir
mv: typeof shelljs.mv
sed: typeof shelljs.sed
tempdir: typeof shelljs.tempdir
test: typeof shelljs.test
which: typeof shelljs.which
rm: typeof trashType //use `trash` instead
spawn: typeof child_process.spawn
spawnSync: typeof child_process.spawnSync
fork: typeof child_process.fork

get: AxiosInstance["get"]
put: AxiosInstance["put"]
post: AxiosInstance["post"]
patch: AxiosInstance["patch"]

readFile: typeof fsPromises.readFile
writeFile: typeof fsPromises.writeFile
appendFile: typeof fsPromises.appendFile
createWriteStream: typeof fs.createWriteStream
readdir: typeof fsPromises.readdir

compile: typeof handlebars.compile_

cwd: typeof process.cwd
pid: typeof process.pid
stderr: typeof process.stderr
stdin: typeof process.stdin
stdout: typeof process.stdout
uptime: typeof process.uptime

path: typeof import("path")

_: LoDashStatic

uuid: typeof uuidType.v4
chalk: ChalkFunction
paste: typeof clipboardy.read
copy: typeof clipboardy.write

db: DB

trash: typeof trashType

wait: Wait

home: Path
isFile: IsCheck
isDir: IsCheck
isBin: IsCheck

tmp: Path
inspect: Inspect

compileTemplate: CompileTemplate

md: Markdown

applescript: AppleScript

setPlaceholder: SetAppProp
setPanel: SetAppProp
setHint: SetAppProp
setInput: SetAppProp
setIgnoreBlur: SetAppProp
setChoices: (choices: Choices<any>) => void

sendResponse: (value: any) => void


show: ShowAppWindow
showImage: ShowAppWindow

edit: Edit

terminal: (script: string) => Promise<string>
iterm: (iterm: string) => Promise<string>


notify: typeof notifyType
```

**Comment by theianjones** [comment]
This is really helpful for me to know everything thats available.

**Reply by johnlindquist** [reply]
I've considered making a `kit` global with all of the global apis built on it so autocomplete would be more obvious (autocomplete works now, these just get lost in the trees).

I really like the look/feel of the scripts without the `kit` global, but I also can see how much it would help. 🤔

**Reply by zacjones93** [reply]
A consistent autocomplete would be really helpful for me too. Could you have both?

**Reply by johnlindquist** [reply]
The autocomplete for the globals is as good as it can be now. (kit is an npm package with a kit.d.ts that code, etc reads in) The apis are all just so short, it's sometimes non-obvious.

1. Are you seeing an inconsistent behavior there?

2. Would you prefer to write scripts in TypeScript? It's very possible, I'd just have to set up a convention like a "ts" dir in the kenv

**Reply by JoepKockelkorn** [reply]
Writing scripts in TS would be very nice, it would feel like home

**Reply by abernier** [reply]
wouldn't be both possible:
- globals:
  ```js
  import "@johnlindquist/kit"
  div("I like globals")
  ```
- namespace:
  ```js
  import * as kit from "@johnlindquist/kit"
  kit.div("I like namespaces")
  ```
- named-import:
  ```js
  import { div } from "@johnlindquist/kit"
  div("I like named-imports")
  ```

**Comment by gr2m** [comment]
`arg` is missing I think?

I wonder if there is a command I could use to show a message to the user for a few seconds? My script result is setting the user's clipboard, it would be nice UX to see a message that it worked

**Reply by johnlindquist** [reply]
This group let's you control the prompt manually:
```js
setPlaceholder: SetAppProp
setPanel: SetAppProp
setHint: SetAppProp
setInput: SetAppProp
setIgnoreBlur: SetAppProp
setChoices: (choices: Choices<any>) => void
```

This is by far the most common:

```js
setPlaceholder(`Copied to clipboard`)
await wait(2000)
```

I may need to abstract it away since it's used so much.

**Reply by gr2m** [reply]
Yeah I made it work with this

```js
setPlaceholder(`"${userFirstNameOrUsername}" copied to your clipboard`);
setIgnoreBlur(false);
setHint(`Press esc to close this prompt`);
```

Is there a way to explicitly close the prompt? Right now it doesn't unless I blur it

**Reply by johnlindquist** [reply]
Three lines of thought:

1. Close prompt
Ends the currently running script. (Or represents the currently running script ending/erroring)

2. Hide prompt

Consider this "internal" for now, but this:

```js
send("HIDE_APP");
```

Needs to be done when sending keystrokes or other scripts that give focus to other apps, but you still want the script running.y

3. Ignore blur

You want the prompt to stay open (used by default in `await drop()` so a user can grab text/images/whatever to put into the prompt without closing it.

**Reply by gr2m** [reply]
I just figured out a workaround that works for me

```js
// wait 5s then exit
await wait(5000);
exit(0);
```

I guess `exit(0)` is the equivalent of ending the running script?

**Reply by johnlindquist** [reply]
Yup, `exit()` is the shelljs `exit` which is literally an alias for [process.exit](https://github.com/shelljs/shelljs/blob/master/shell.js#L32). So I guess I'm aliasing and alias 😅

**Comment by SevenOutman** [comment]
Could this list of methods be categorized by purposes, e.g. Inputs, Outputs? It's not easy for me to find out which command I can use to display a piece of message (outputting).
